File: app/src/main/java/com/sygic/travel/sdkdemo/ReferenceActivity.java
package com.sygic.travel.sdkdemo;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.Menu;
import android.view.MotionEvent;
import android.view.View;
import android.webkit.WebChromeClient;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.ProgressBar;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

import static com.sygic.travel.sdk.model.place.Reference.TITLE;
import static com.sygic.travel.sdk.model.place.Reference.URL;

public class ReferenceActivity extends AppCompatActivity {

	private WebView webView;
	private ProgressBar progressBar;

	private boolean loaded = false;
	private boolean touch = false;

	private List<String> loadedUrls = new ArrayList<>();
	private String referenceUrl;
	private String referenceTitle;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_reference);

		if(getIntent().getExtras() != null) {
			referenceUrl = getIntent().getExtras().getString(URL, "");
			referenceTitle = getIntent().getExtras().getString(TITLE, getString(R.string.title_activity_reference));
		}
		setTitle(referenceTitle);
		initWebView();
		loadReferenceUrl();
	}

	private void initWebView() {
		if (loaded) {
			return;
		}
		progressBar = (ProgressBar) findViewById(R.id.pb_reference);

		webView = (WebView) findViewById(R.id.wv_reference);
		webView.getSettings().setJavaScriptEnabled(true);
		webView.getSettings().setDomStorageEnabled(true);
		webView.getSettings().setSupportZoom(true);
		webView.getSettings().setBuiltInZoomControls(true);
		webView.getSettings().setDisplayZoomControls(false);
		webView.setWebViewClient(getOnUrlOverrideClient());
		webView.setWebChromeClient(getOnConsoleMessageClient());
		webView.getSettings().setLoadWithOverviewMode(true);
		webView.getSettings().setUseWideViewPort(true);

		webView.setOnTouchListener(getOnTouchListener());

		webView.setWebChromeClient(new WebChromeClient() {
			public void onProgressChanged(WebView view, int progress) {
				if(progress < 100 && progressBar.getVisibility() == ProgressBar.GONE){
					progressBar.setVisibility(ProgressBar.VISIBLE);
				}

				progressBar.setProgress(progress);
				if(progress == 100) {
					progressBar.setVisibility(ProgressBar.GONE);
				}
			}
		});
	}

	private void loadReferenceUrl() {
		webView.loadUrl(referenceUrl);
		loaded = true;
		loadedUrls.add(referenceUrl);
	}

	public WebChromeClient getOnConsoleMessageClient(){
		return new WebChromeClient() {
			@Override
			public void onProgressChanged(WebView view, int progress) {
				ReferenceActivity.this.setProgress(progress * 1000);
			}
		};
	}

	private View.OnTouchListener getOnTouchListener() {
		return new View.OnTouchListener() {
			@Override
			public boolean onTouch(View v, MotionEvent event) {
				touch = true;
				return false;
			}
		};
	}

	private WebViewClient getOnUrlOverrideClient() {
		return new WebViewClient(){
			@Override
			public boolean shouldOverrideUrlLoading(WebView view, String url) {
				if(touch){
					try {
						loadedUrls.add(url);
						webView.loadUrl(url);
						return true;
					} catch(Throwable throwable){
						ReferenceActivity.this.finish();
						return false;
					}
				} else {
					return false;
				}
			}
		};
	}
}

File: app/src/main/java/com/sygic/travel/sdkdemo/detail/ReferenceActivity.kt
package com.sygic.travel.sdkdemo.detail

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.webkit.WebChromeClient
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.ProgressBar
import com.sygic.travel.sdkdemo.R
import java.util.ArrayList

class ReferenceActivity : AppCompatActivity() {

	private var webView: WebView? = null
	private var progressBar: ProgressBar? = null

	private var loaded = false
	private var touch = false

	private val loadedUrls = ArrayList<String?>()
	private var referenceUrl: String? = null
	private var referenceTitle: String? = null

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_reference)

		// Reference is basically only URL to be loaded in a web view
		if (intent.extras != null) {
			referenceUrl = intent.extras.getString(REFERENCE_URL, "")
			referenceTitle = intent.extras.getString(REFERENCE_TITLE, getString(R.string.title_activity_reference))
		}
		title = referenceTitle
		initWebView()
		loadReferenceUrl()
	}

	// Web view initialization
	private fun initWebView() {
		if (loaded) {
			return
		}
		progressBar = findViewById(R.id.pb_reference)

		webView = findViewById(R.id.wv_reference)
		webView!!.settings.javaScriptEnabled = true
		webView!!.settings.domStorageEnabled = true
		webView!!.settings.setSupportZoom(true)
		webView!!.settings.builtInZoomControls = true
		webView!!.settings.displayZoomControls = false
		webView!!.setWebViewClient(onUrlOverrideClient)
		webView!!.setWebChromeClient(onConsoleMessageClient)
		webView!!.settings.loadWithOverviewMode = true
		webView!!.settings.useWideViewPort = true

		webView!!.setOnTouchListener(onTouchListener)

		webView!!.setWebChromeClient(object : WebChromeClient() {
			override fun onProgressChanged(view: WebView, progress: Int) {
				if (progress < 100 && progressBar!!.visibility == ProgressBar.GONE) {
					progressBar!!.visibility = ProgressBar.VISIBLE
				}

				progressBar!!.progress = progress
				if (progress == 100) {
					progressBar!!.visibility = ProgressBar.GONE
				}
			}
		})
	}

	private fun loadReferenceUrl() {
		webView!!.loadUrl(referenceUrl)
		loaded = true
		loadedUrls.add(referenceUrl)
	}

	val onConsoleMessageClient: WebChromeClient
		get() = object : WebChromeClient() {
			override fun onProgressChanged(view: WebView, progress: Int) {
				this@ReferenceActivity.setProgress(progress * 1000)
			}
		}

	private val onTouchListener: View.OnTouchListener
		get() = View.OnTouchListener { _, _ ->
			touch = true
			false
		}

	// Since we support Android API versions 15 and newer, we use the deprecated shouldOverrideUrlLoading method
	private val onUrlOverrideClient: WebViewClient
		get() = object : WebViewClient() {
			override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
				if (touch) {
					try {
						loadedUrls.add(url)
						webView!!.loadUrl(url)
						return true
					} catch (throwable: Throwable) {
						this@ReferenceActivity.finish()
						return false
					}

				} else {
					return false
				}
			}
		}

	companion object {
		val REFERENCE_TITLE = "reference_title"
		val REFERENCE_URL = "reference_url"
	}
}
